### PR TITLE
Add Android ability to auto-close browser after returning to the app

### PR DESCRIPTION
**Pull request by Hiilbig**
On Android the login browser is kept open in the background when the authentication window return control to the main application. If the customTabsIntent is called with the flags ActivityFlags.NoHistory and ActivityFlags.ClearTop, the browser will close itself. See https://forums.xamarin.com/discussion/100286/xamarin-auth-android-chrome-custom-tabs-doesnt-close-on-redirect for further details

### Changes

Please describe both what is changing and why this is important. Include:

- Added property AutoCloseBrowser to class AndroidBrowserBase
- Added optional parameter to the AndroidBrowserBase to initialize the AutoCloseBrowser property.
- Extended Auth0ClientOptions with an 'AutoCloseBrowser' option to be used in the Auth0Client initialization/creation.

### References

Please include relevant links supporting this change such as a:
[This discussion]( https://forums.xamarin.com/discussion/100286/xamarin-auth-android-chrome-custom-tabs-doesnt-close-on-redirect)

### Testing

I implemented the library manually into my own project and added the new feature and verified the Auth0 browser window indeed was closed after I returned to the main app again.

- [ ] This change adds unit test coverage
- [ ] This change adds integration test coverage
- [x] This change has been tested on the latest version of the platform/language.

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [x] All existing and new tests complete without errors
- [x] All code guidelines in the [CONTRIBUTING documentation](../CONTRIBUTING.md) have been run/followed

### CODE REVIEW

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (frederikprijck) [inline src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:35]: What's the purpose of using `NoHistory` here ? As I understand correctly, `ClearTop` should be the one solving the issue, or is this incorrect?

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (frederikprijck) [inline src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:33]: I understand this is added to guarantee backwards compatibility, but I am curious to consider whether or not this is to be seen as a breaking change or a fix. Not saying this needs to be reverted, but curious to hear your thought as to why you went with the opt-in behavior?

**Review by frederikprijck — CHANGES_REQUESTED** [review]
Thanks for the PR, super apprectaed! Left a few questions.

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (Hiilbig) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:35]: The reason I included the "NoHistory" flag was based on the description found here: https://developer.android.com/guide/topics/manifest/activity-element#nohist
I did not want the Auth0 window to be accessible with the "Back" command in Android. Even if it Auth0 activity had been closed, a new would be created, letting the user come back to the login screen.

The FLAG_ACTIVITY_CLEAR_TOP is used for clearing any other activities on the navigation stack and return navigation to the initial (old) activity and then let it receive the intent.

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (Hiilbig) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:33]: I did not want to assume everybody would want this change, so I provided a way to let the library work "as previously designed", to make sure an update to the library did not break anything for everyone else. Furthermore your code of conduct for any contribution states "no breaking changes" unless discussed with you...

**Comment by Hiilbig** [comment]
This is a suggested fix for issue #202

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (frederikprijck) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:35]: I am curious to understand what the intention is for adding `NoHistory` here. How does it relate to the problem mentioned about the browser not being closed.

What is the actual issue/risk you see with the login page being part of the history?

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (frederikprijck) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:33]: Assuming we only go with ClearTop, I am leaning towards thinking this is more a bug fix as it is a change that is to be considered breaking for the end user. I have a feeling the extra code here is making things more complicated while this probably could/should be the default behavior.

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (Hiilbig) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:35]: By leaving in the history the login page, the users could end up going "back" to the login page and login again, without a proper logout first.
There are some behavior questions that need to be answered for this, like

- What happens with the app if the user does another login?
- What happens with the app if the user login with a different account?
- How do we manage the transition "Back" to the login page? Do we prompt the user? Do we auto-logout? What if the "back" was an accident?

The only way to avoid this easily, is to only be able to logout and then login again - and to make sure one user does not have access to the previous users information, is to clear the navigation history.

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (Hiilbig) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:33]: If you prefer the auto-close functionality to be the default behavior, I can make that change if you want to, either by just forcing the extra flag(s) onto the intent (The least new code), or by defaulting the new property to "true" (the lease change to my suggested fix). What do you prefer?

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (frederikprijck) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:35]: All our browser-based SDKs (https://github.com/auth0/auth0-spa-js, https://github.com/auth0/auth0-angular, https://github.com/auth0/auth0-react) as well as our mobile Java SDK (https://github.com/auth0/Auth0.Android/blob/main/auth0/src/main/java/com/auth0/android/provider/AuthenticationActivity.kt#L101) end up having the login page end up in history, which is why I am not sure we should add it here. 

The problems you describe span all of these SDKs and haven't shown to be an actual issue so far.

The intention here is solving an issue that seems to be unrelated to removing the login page from the browser history, if you believe this is an issue I would recommend opening an issue so we can take this internally to discuss with all of the maintainers of the mentioned SDKs that are impacted by the login page ending up in the browser's history.

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (frederikprijck) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:33]: I will discuss this internally with the maintainer of the Android SDK, who's a bit more familiar with the different flags and whether or not it should have been the default or not. Will get back to this.

File: src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs
Comment (Hiilbig) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:35]: It might be an edge case that no-one else have given much thought. I do not know if this is an issue for Xamarin.droid only, but I do not think it is. I strongly believe there is a genuine security risk for being able to go back to the login screen and basically be able to go "forward" again, not knowing who is logged in.
If the decision is to leave out that part of it, I would strongly recommend a solution where the flag can be added or removed by choice.
I will create a separate issue for this.

Comment (frederikprijck) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:35]: I am happy to raise your concern internally with our security team.

For what I know, once u are logged in with Auth0 you have a session with our Auth0 server. Hitting back visits the login page, identifies u have a session and instantly redirects you. 

Before being able to actually raise it internally, I would like to gather some more information about the security risk you are seeing. Would you happen to want to reflect your concern in the created issue (https://github.com/auth0/auth0-oidc-client-net/issues/204) rather than mentioning you where asked to create an issue ?  This will help us driving the conversation with the security team.

Thanks 🙏

Comment (Hiilbig) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:35]: You are welcome, I appreciate you appreciating it!

This library is great and I am happy to be to continue to use it - and if I can help making it better in any way, even if it is just for the sake of the argument and making sure the concern has been addressed and given some thought. I could easily be wrong in my assessment of the risk, but we can't know until we have talked about it.

I will remove the "NoHistory" flag from the PR and then leave it up to issue #204 to have the other part addressed, if needed.

**Did you want me to keep the backward compatibility (extra code added to keep the default behavior as it is now) or do you want me to just add the extra flag (ActivityFlags.ClearTop) to close the browser (and then change the default behavior of the library)?**

Comment (frederikprijck) [reply src/Auth0.OidcClient.Android/ChromeCustomTabsBrowser.cs:35]: > Did you want me to keep the backward compatibility (extra code added to keep the default behavior as it is now) or do you want me to just add the extra flag (ActivityFlags.ClearTop) to close the browser (and then change the default behavior of the library)?

Will get back to this asap, sorry about that delay. Feel free to keep the code as is for now.

**Comment by frederikprijck** [comment]
Hi @Hiilbig ,

Could you elaborate on the actual situation that causes the problem for u? As well as some extra contextual information?

- Android version
- Browser used

If possible a screen recording could help. I was looking into this myself and came to the following conclusion:

- Chrome Custom Tabs without ClearTop: Browser closes
- Chrome Custom Tabs with ClearTop: Browser closes

- Opera Without ClearTop: Browser doesn't close
- Opera with ClearTop: Browser doesn't close

All of these results are expected and indicate that adding `ClearTop` does not change anything. However, I assume it can depend on the android version being used, so would love to see if we can get the behavior reproduced and verify the fix.

Note that in the case of using a None-Chrome Custom Tab browser, we are unable to close the browser and it will keep running in the background, which is expected. 

Thanks

**Pull request closed without merging**
